README.md
[![Build Status](https://secure.travis-ci.org/alexfernandez/loadtest.png)](http://travis-ci.org/alexfernandez/loadtest)

# loadtest

Runs a load test on the selected HTTP URL. The API allows for easy integration in your own tests.

Why use `loadtest` instead of any other of the available tools, notably Apache `ab`?
`loadtest` allows you to configure and tweak requests to simulate real world loads.

The set of basic options are designed to be compatible with Apache `ab`.
But while `ab` can only set a concurrency level and lets the server adjust to it,
`loadtest` allows you to set a rate or requests per second with the `--rps` option.
Example:

    loadtest -c 10 --rps 200 http://mysite.com/

This command sends exactly 200 requests per second with concurrency 10,
so you can see how your server copes with sustained rps.
Even if `ab` reported a rate of 200 rps,
you will be surprised to see how a constant rate of requests per second affects performance:
no longer are the requests adjusted to the server, but the server must adjust to the requests!
Rps rates are usually lowered dramatically about 20~25% (in our example from 200 to 150 rps),
but the resulting figure is much more robust.

Using the provided API it is very easy to integrate loadtest with your package, and run programmatic load tests.
loadtest makes it very easy to run load tests as part of systems tests, before deploying a new version of your software.
The results include mean response times and percentiles,
so that you can abort deployment e.g. if 99% of the requests don't finish in 10 ms or less.

## Change Log

Latest significant changes.

### Changes in Version 1.1

* Mechanism to generate different POST and PUT bodies using a function.
* Duplicate headers are now ignored; set them using `-H header:value1;value2`.

## Changes in Version 1.0

* Option parsing has been improved; no longer is a `true` needed after certain options.
* Requests per second specified with `--rps` are now total rps, instead of multiplied by concurrency.
* Option `--agent` has been deprecated in favor of `--keepalive`.
* Support for Node.js < 0.10 removed.

## Installation

Install globally as root:

    # npm install -g loadtest

On Ubuntu or Mac OS X systems install using sudo:

    $ sudo npm install -g loadtest

For access to the API just add package `loadtest` to your `package.json` devDependencies:

    {
        ...
        "devDependencies": {
            "loadtest": "*"
        },
        ...
    }

## Basic Usage

Run as a script to load test a URL:

    $ loadtest [-n requests] [-c concurrency] URL

The URL can be "http://" or "https://". Set the max number of requests with `-n`, and the desired level of concurrency with the `-c` parameter.

Single-dash parameters (e.g. `-n`) are designed to be compatible with Apache `ab`.
  http://httpd.apache.org/docs/2.2/programs/ab.html

To get online help, run loadtest without parameters:

    $ loadtest

### Regular Usage

The following parameters are compatible with Apache ab.

#### `-n requests`

Number of requests to send out.

Note: the total number of requests sent can be bigger than the parameter, if there is a concurrency parameter
and number of requests is not a multiple of concurrency.

#### `-c concurrency`

loadtest will create a simultaneous number of clients; this parameter controls how many.

#### `-t timelimit`

Max number of seconds to wait until requests no longer go out.

Note: this is different than Apache `ab`, which stops _receiving_ requests after the given seconds.

### `-C cookie-name=value`

Send a cookie with the request. The cookie `name=value` is then sent to the server.
This parameter can be repeated as many times as needed.

### `-H header:value`

Send a custom header with the request. The line `header:value` is then sent to the server.
This parameter can be repeated as many times as needed.
Example:

    $ loadtest -H user-agent:tester/0.4 ...

Note: if not present, loadtest will add a few headers on its own: the "host" header parsed from the URL,
a custom user agent "loadtest/" plus version (`loadtest/1.1.0`), and an accept header for "\*/\*".

Note: when the same header is sent several times, only the last value will be considered.
If you want to send multiple values with a header, separate them with semicolons:

    $ loadtest -H accept:text/plain;text-html ...

### `-T content-type`

Set the MIME content type for POST data. Default: `text/plain`.

### `-P POST-body`

Send the string as the POST body. E.g.: `-P '{"key": "a9acf03f"}'`

### `-p POST-file`

Send the data contained in the given file in the POST body.
Remember to set `-T` to the correct content-type.  

If `POST-file` has `.js` extension it will be `require`d. It should be a valid node module and it
should `export` a single function, which is invoked with an automatically generated request identifier
to provide the body of each request.
This is useful if you want to generate request bodies dynamically and vary them for each request.

Example:

    module.exports = function(requestId) {
      // this object will be serialized to JSON and sent in the body of the request
      return {
        key: 'value',
        requestId: requestId
      };
    };

### `-u PUT-file`

Send the data contained in the given file as a PUT request.
Remember to set `-T` to the correct content-type.  

If `PUT-file` has `.js` extension it will be `require`d. It should be a valid node module and it
should `export` a single function, which is invoked with an automatically generated request identifier
to provide the body of each request.
This is useful if you want to generate request bodies dynamically and vary them for each request.
For an example function see above for `-p`.

#### `-r`

Recover from errors. Always active: loadtest does not stop on errors.
After the tests are finished, if there were errors a report with all error codes will be shown.

#### `-V`

Show version number and exit.

### Advanced Usage

The following parameters are _not_ compatible with Apache ab.

#### `--rps requestsPerSecond`

Controls the number of requests per second that are sent.
Can be fractional, e.g. `--rps 0.5` sends one request every two seconds.

Note: Concurrency doesn't affect the final number of requests per second,
since rps will be shared by all the clients. E.g.:

    loadtest <url> -c 10 --rps 10

will send a total of 10 rps to the given URL, from 10 different clients
(each client will send 1 request per second).

Beware: if concurrency is too low then it is possible that there will not be enough clients
to send all of the rps, adjust it with `-c` if needed.

#### `--keepalive`

Open connections using keep-alive: use header 'Connection: Keep-alive' instead of 'Connection: Close'.

Note: Uses [agentkeepalive](https://npmjs.org/package/agentkeepalive),
which performs better than the default node.js agent.

#### `--agent` (deprecated)

Open connections using keep-alive.

Note: instead of using the default agent, this option is now an alias for `--keepalive`.

#### `--quiet`

Do not show any messages.

#### `--debug`

Show debug messages.

#### `--insecure`

Allow invalid and self-signed certificates over https.

### Server

loadtest bundles a test server. To run it:

    $ testserver [--delay ms] [port]

This command will show the number of requests received per second,
the latency in answering requests and the headers for selected requests.

The server returns a short text 'OK' for every request,
so that latency measurements don't have to take into account request processing.

If no port is given then default port 7357 will be used.
The optional delay instructs the server to wait for the given number of milliseconds
before answering each request, to simulate a busy server.

### Complete Example

Let us now see how to measure the performance of the test server.

First we install `loadtest` globally:

    $ sudo npm install -g loadtest

Now we start the test server:

    $ testserver
    Listening on port 7357

On a different console window we run a load test against it for 20 seconds
with concurrency 10 (only relevant results are shown):

    $ loadtest http://localhost:7357/ -t 20 -c 10
    ...
    Requests: 9589, requests per second: 1915, mean latency: 10 ms
    Requests: 16375, requests per second: 1359, mean latency: 10 ms
    Requests: 16375, requests per second: 0, mean latency: 0 ms
    ...
    Completed requests:  16376
    Requests per second: 368
    Total time:          44.503181166000005 s
    
    Percentage of the requests served within a certain time
      50%      4 ms
      90%      5 ms
      95%      6 ms
      99%      14 ms
     100%      35997 ms (longest request)

Results were quite erratic, with some requests taking up to 36 seconds;
this suggests that Node.js is queueing some requests for a long time, and answering them irregularly.
Now we will try a fixed rate of 1000 rps:

    $ loadtest http://localhost:7357/ -t 20 -c 10 --rps 1000
    ...
    Requests: 4551, requests per second: 910, mean latency: 0 ms
    Requests: 9546, requests per second: 1000, mean latency: 0 ms
    Requests: 14549, requests per second: 1000, mean latency: 20 ms
    ...
    Percentage of the requests served within a certain time
      50%      1 ms
      90%      2 ms
      95%      8 ms
      99%      133 ms
     100%      1246 ms (longest request)

Again erratic results. In fact if we leave the test running for 50 seconds we start seeing errors:

    $ loadtest http://localhost:7357/ -t 50 -c 10 --rps 1000
    ...
    Requests: 29212, requests per second: 496, mean latency: 14500 ms
    Errors: 426, accumulated errors: 428, 1.5% of total requests

Let us lower the rate to 500 rps:

    $ loadtest http://localhost:7357/ -t 20 -c 10 --rps 5000
    ...
    Requests: 0, requests per second: 0, mean latency: 0 ms
    Requests: 2258, requests per second: 452, mean latency: 0 ms
    Requests: 4757, requests per second: 500, mean latency: 0 ms
    Requests: 7258, requests per second: 500, mean latency: 0 ms
    Requests: 9757, requests per second: 500, mean latency: 0 ms
    ...
    Requests per second: 500
    Completed requests:  9758
    Total errors:        0
    Total time:          20.002735398000002 s
    Requests per second: 488
    Total time:          20.002735398000002 s
    
    Percentage of the requests served within a certain time
      50%      1 ms
      90%      1 ms
      95%      1 ms
      99%      14 ms
     100%      148 ms (longest request)

Much better: a sustained rate of 500 rps is seen most of the time,
488 rps average, and 99% of requests answered within 14 ms.

We now know that our server can accept 500 rps without problems.
Not bad for a single-process naïve Node.js server...
We may refine our results further to find at which point from 500 to 1000 rps our server breaks down.

## API

`loadtest` is not limited to running from the command line; it can be controlled using an API,
thus allowing you to load test your application in your own tests.

### Invoke Load Test

To run a load test, just call the exported function `loadTest()` with a set of options and an optional callback:

    var loadtest = require('loadtest');
    var options = {
        url: 'http://localhost:8000',
        maxRequests: 1000,
    };
    loadtest.loadTest(options, function(error, result)
    {
        if (error)
        {
            return console.error('Got an error: %s', error);
        }
        console.log('Tests run successfully');
    });

The callback `function(error, result)` will be invoked when the max number of requests is reached,
or when the max number of seconds has elapsed.

Beware: if there are no `maxRequests` and no `maxSeconds`, then tests will run forever
and will not call the callback.

### Options

All options but `url` are, as their name implies, optional.

#### `url`

The URL to invoke. Mandatory.

#### `concurrency`

How many clients to start in parallel.

#### `maxRequests`

A max number of requests; after they are reached the test will end.

#### `maxSeconds`

Max number of seconds to run the tests.

Note: after the given number of seconds `loadtest` will stop sending requests,
but may continue receiving tests afterwards.

#### `cookies`

An array of cookies to send. Each cookie should be a string of the form name=value.

#### `headers`

A map of headers. Each header should be an entry in the map with the value given as a string.
If you want to have several values for a header, write a single value separated by semicolons,
like this:

    {
        accept: "text/plain;text/html"
    }

Note: when using the API, the "host" header is not inferred from the URL but needs to be sent
explicitly.

#### `method`

The method to use: POST, PUT. Default: GET.

#### `body`

The contents to send in the body of the message, for POST or PUT requests.
Can be a string or an object (which will be converted to JSON).

#### `contentType`

The MIME type to use for the body. Default content type is `text/plain`.

#### `requestsPerSecond`

How many requests each client will send per second.

#### `agentKeepAlive`

Use an agent with 'Connection: Keep-alive'.

Note: Uses [agentkeepalive](https://npmjs.org/package/agentkeepalive),
which performs better than the default node.js agent.

#### `quiet`

Do not show any messages.

#### `indexParam`

The given string will be replaced in the final URL with a unique index.
E.g.: if URL is `http://test.com/value` and `indexParam=value`, then the URL
will be:
* http://test.com/1
* http://test.com/2
* ...

#### `insecure`

Allow invalid and self-signed certificates over https.

### Results

The results passed to your callback at the end of the load test contains a full set of data, including:
mean latency, number of errors and percentiles.
An example follows:

    {
      totalRequests: 1000,
      percentiles: {
        '50': 7,
        '90': 10,
        '95': 11,
        '99': 15
      },
      rps: 2824,
      totalTimeSeconds: 0.354108,
      meanLatencyMs: 7.72,
      totalErrors: 3,
      errors: {
        '0': 1,
        '500': 2
      }
    }

### Start Test Server

To start the test server use the exported function `startServer()` with a set of options and an optional callback:

    var testserver = require('testserver');
    var server = testserver.startServer({ port: 8000 });

This function returns an HTTP server which can be `close()`d when it is no longer useful.

The following options are available.

#### `port`

Optional port to use for the server.

Note: the default port is 7357, since port 80 requires special privileges.

#### `delay`

Wait the given number of milliseconds to answer each request.

### Complete Example

The file `lib/sample.js` shows a complete example, which is also a full integration test:
it starts the server, send 1000 requests, waits for the callback and closes down the server.

## License

(The MIT License)

Copyright (c) 2013-4 Alex Fernández <alexfernandeznpm@gmail.com>
and [contributors](https://github.com/alexfernandez/loadtest/graphs/contributors).

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the 'Software'), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.



CHANGELOG.md
# loadtest Changelog

This file tracks only versions from 0.2.0 to 0.2.8, and is kept here only for historic purposes.

## Version History

Release date, version number and changes.

### 2013-08-30: version 0.2.0

* Added option -C to send cookies.
* Added options -T, -p and -u to send POST and PUT requests.

### 2013-09-01: version 0.2.1

* Added bin/ files to avoid cross-platform issues.
  Parse arguments using `optimist`.
  Pull request #9 by https://github.com/zaphod1984
* Use native process.hrtime() instead of Microtime.
  Pull request #8 by to https://github.com/zaphod1984
* Changed binary testserver to testserver-loadtest, to avoid conflicts.
  Issue #4. Thanks to https://github.com/zaphod1984

### 2013-09-02: version 0.2.2

* Emergency fix for binary paths.
  Pull request #10 by to https://github.com/zaphod1984
* Fix library paths.
* Added #env lines to binary paths to run on Unix-like systems.

### 2013-09-06: version 0.2.3

* Update sample.js: fix option key for POST body.
  Pull request #11 by https://github.com/danieltdt
* Fix maxSeconds option, added test.
  Pull request #12 by https://github.com/danieltdt
* Show more complete information: concurrency, agent...
* Clean up for loadtest.run(), should not be used.

### 2013-09-12: version 0.2.4

* Remove cloneextend since it is not used.
  Pull request #13 by https://github.com/zaphod1984
* Start default test server when invoking lib/testserver.js.
* Change default port for test server to 7357.

### 2013-09-28: version 0.2.5

* JSHint clean up.
* Option --index should be working again.

### 2013-09-30: version 0.2.6

* Use the deployment module.
* Correct support for https.

### 2014-01-16: version 0.2.7

* Corrected bug in -C option for sending cookies.
* Set the correct Cookie header for test cookies.

### 2014-01-28: version 0.2.8.

* Issue 15: Added option -H to send a custom header.
* Separate multiple cookies with a semicolon instead of a comma: `cookie1=1; cookie2=2`.



lib/timing.js
'use strict';

/**
 * Measure latency for a load test.
 * (C) 2013 Alex Fernández.
 */


// requires
require('prototypes');
var testing = require('testing');
var util = require('util');
var crypto = require('crypto');
var Log = require('log');

// globals
var log = new Log('info');


/**
 * Latency measurements. Options can be:
 *	- maxRequests: max number of requests to measure before stopping.
 *	- maxSeconds: max seconds, alternative to max requests.
 *	- quiet: do not log messages.
 * An optional callback(error, results) will be called with an error,
 * or the results after max is reached.
 */
exports.Latency = function(options, callback)
{
	// self-reference
	var self = this;

	// attributes
	var requests = {};
	var partialRequests = 0;
	var partialTime = 0;
	var partialErrors = 0;
	var lastShown = getTime();
	var initialTime = getTime();
	var totalRequests = 0;
	var totalTime = 0;
	var totalErrors = 0;
	var maxLatencyMs = 0;
	var histogramMs = {};
	var errorCodes = {};
	var running = true;

	// init
	if (options.quiet)
	{
		log.level = Log.NOTICE;
	}
	if (options.debug)
	{
		log.level = Log.DEBUG;
	}

	/**
	 * Start the request with the given id.
	 */
	self.start = function(requestId)
	{
		requestId = requestId || createId();
		requests[requestId] = getTime();
		return requestId;
	};

	/**
	 * Compute elapsed time and add the measurement.
	 * Accepts an optional error code signaling an error.
	 */
	self.end = function(requestId, errorCode)
	{
		if (!(requestId in requests))
		{
			log.error('Message id ' + requestId + ' not found');
			return;
		}
		if (!running)
		{
			return;
		}
		add(getElapsed(requests[requestId]), errorCode);
		delete requests[requestId];
	};

	/**
	 * Add a new measurement, possibly removing an old one.
	 * Accepts an optional error code signaling an error.
	 */
	function add(time, errorCode)
	{
		log.debug('New value: %s', time);
		partialTime += time;
		partialRequests++;
		totalTime += time;
		totalRequests++;
		if (errorCode)
		{
			errorCode = '' + errorCode;
			partialErrors++;
			totalErrors++;
			if (!(errorCode in errorCodes))
			{
				errorCodes[errorCode] = 0;
			}
			errorCodes[errorCode] += 1;
		}
		log.debug('Partial requests: %s', partialRequests);
		var rounded = Math.floor(time);
		if (rounded > maxLatencyMs)
		{
			maxLatencyMs = rounded;
		}
		if (!histogramMs[rounded])
		{
			log.debug('Initializing histogram for %s', rounded);
			histogramMs[rounded] = 0;
		}
		histogramMs[rounded] += 1;
		if (isFinished())
		{
			finish();
		}
	}

	/**
	 * Show latency for partial requests.
	 */
	self.showPartial = function()
	{
		var elapsedSeconds = getElapsed(lastShown) / 1000;
		var meanTime = partialTime / partialRequests || 0.0;
		var results = {
			meanLatencyMs: Math.round(meanTime / 10) * 10,
			rps: Math.round(partialRequests / elapsedSeconds)
		};
		var percent = '';
		if (options.maxRequests)
		{
			percent = ' (' + Math.round(100 * totalRequests / options.maxRequests) + '%)';
		}
		log.info('Requests: %s%s, requests per second: %s, mean latency: %s ms', totalRequests, percent, results.rps, results.meanLatencyMs);
		if (totalErrors)
		{
			percent = Math.round(100 * 10 * totalErrors / totalRequests) / 10;
			log.info('Errors: %s, accumulated errors: %s, %s% of total requests', partialErrors, totalErrors, percent);
		}
		partialTime = 0;
		partialRequests = 0;
		partialErrors = 0;
		lastShown = getTime();
	};

	/**
	 * Returns the current high-resolution real time in a [seconds, nanoseconds] tuple Array
	 * @return {*}
	 */
	function getTime() {
		return process.hrtime();
	}

	/**
	 * calculates the elapsed time between the assigned startTime and now
	 * @param startTime
	 * @return {Number} the elapsed time in milliseconds
	 */
	function getElapsed(startTime) {
		var elapsed = process.hrtime(startTime);
		return elapsed[0] * 1000 + elapsed[1] / 1000000;
	}

	/**
	 * Check out if the measures are finished.
	 */
	function isFinished()
	{
		log.debug('Total requests %s, max requests: %s', totalRequests, options.maxRequests);
		if (options.maxRequests && totalRequests >= options.maxRequests)
		{
			log.debug('Max requests reached: %s', totalRequests);
			return true;
		}
		var elapsedSeconds = getElapsed(initialTime) / 1000;
		if (options.maxSeconds && elapsedSeconds >= options.maxSeconds)
		{
			log.debug('Max seconds reached: %s', totalRequests);
			return true;
		}
		return false;
	}

	/**
	 * We are finished.
	 */
	function finish()
	{
		running = false;
		if (callback)
		{
			return callback(null, self.getResults());
		}
		self.show();
	}

	/**
	 * Get final results.
	 */
	self.getResults = function()
	{
		var elapsedSeconds = getElapsed(initialTime) / 1000;
		var meanTime = totalTime / totalRequests;
		return {
			totalRequests: totalRequests,
			totalErrors: totalErrors,
			totalTimeSeconds: elapsedSeconds,
			rps: Math.round(totalRequests / elapsedSeconds),
			meanLatencyMs: Math.round(meanTime / 10) * 10,
			percentiles: self.computePercentiles(),
			errorCodes: errorCodes
		};
	};

	/**
	 * Compute the percentiles.
	 */
	self.computePercentiles = function()
	{
		log.debug('Histogram: %s', util.inspect(histogramMs));
		var percentiles = {
			50: false,
			90: false,
			95: false,
			99: false
		};
		var counted = 0;

		for (var ms = 0; ms <= maxLatencyMs; ms++)
		{
			if (!histogramMs[ms])
			{
				continue;
			}
			log.debug('Histogram for %s: %s', ms, histogramMs[ms]);
			counted += histogramMs[ms];
			var percent = counted / totalRequests * 100;

			Object.keys(percentiles).forEach(function(percentile) {
				log.debug('Checking percentile %s for %s', percentile, percent);
				if (!percentiles[percentile] && percent > percentile)
				{
					percentiles[percentile] = ms;
				}
			});
		}
		return percentiles;
	};

	/**
	 * Show final results.
	 */
	self.show = function()
	{
		var results = self.getResults();
		log.info('');
		log.info('Target URL:          %s', options.url);
		if (options.maxRequests)
		{
			log.info('Max requests:        %s', options.maxRequests);
		}
		else if (options.maxSeconds)
		{
			log.info('Max time (s):        %s', options.maxSeconds);
		}
		log.info('Concurrency level:   %s', options.concurrency);
		var agent = 'none';
		if (options.agentKeepAlive)
		{
			agent = 'keepalive';
		}
		log.info('Agent:               %s', agent);
		if (options.requestsPerSecond)
		{
			log.info('Requests per second: %s', options.requestsPerSecond * options.concurrency);
		}
		log.info('');
		log.info('Completed requests:  %s', results.totalRequests);
		log.info('Total errors:        %s', results.totalErrors);
		log.info('Total time:          %s s', results.totalTimeSeconds);
		log.info('Requests per second: %s', results.rps);
		log.info('Total time:          %s s', results.totalTimeSeconds);
		log.info('');
		log.info('Percentage of the requests served within a certain time');

		Object.keys(results.percentiles).forEach(function(percentile) {
			log.info('  %s%      %s ms', percentile, results.percentiles[percentile]);
		});

		log.info(' 100%      %s ms (longest request)', maxLatencyMs);
		if (results.totalErrors)
		{
			log.info('');
			log.info(' 100%      %s ms (longest request)', maxLatencyMs);
			log.info('');
			Object.keys(results.errorCodes).forEach(function(errorCode)
			{
				var padding = ' '.repeat(4 - errorCode.length);
				log.info(' %s%s:   %s errors', padding, errorCode, results.errorCodes[errorCode]);
			});
		}
	};
};


/**
 * Create a unique, random token.
 */
function createId()
{
	var value = '' + Date.now() + Math.random();
	var hash = crypto.createHash('sha256');
	return hash.update(value).digest('hex').toLowerCase();
}

/**
 * Test latency ids.
 */
function testLatencyIds(callback)
{
	var latency = new exports.Latency({});
	var firstId = latency.start();
	testing.assert(firstId, 'Invalid first latency id', callback);
	var secondId = latency.start();
	testing.assert(secondId, 'Invalid second latency id', callback);
	testing.assert(firstId != secondId, 'Repeated latency ids', callback);
	testing.success(callback);
}

/**
 * Test latency measurements.
 */
function testLatencyRequests(callback)
{
	var options = {
		maxRequests: 10,
	};
	var errorCode = '500';
	var latency = new exports.Latency(options, function(error, result)
	{
		testing.check(error, 'Could not compute latency', callback);
		testing.assertEquals(result.totalRequests, 10, 'Invalid total requests', callback);
		testing.assertEquals(result.totalErrors, 1, 'Invalid total errors', callback);
		testing.assert(errorCode in result.errorCodes, 'Error code not found', callback);
		testing.assertEquals(result.errorCodes[errorCode], 1, 'Should have one ' + errorCode, callback);
		testing.success(callback);
	});
	var id;
	for (var i = 0; i < 9; i++)
	{
		id = latency.start();
		latency.end(id);
	}
	id = latency.start();
	latency.end(id, errorCode);
}

/**
 * Check that percentiles are correct.
 */
function testLatencyPercentiles(callback)
{
	var options = {
		maxRequests: 10
	};
	var latency = new exports.Latency(options, function(error)
	{
		testing.check(error, 'Error while testing latency percentiles', callback);
		var percentiles = latency.getResults().percentiles;

		Object.keys(percentiles).forEach(function(percentile) {
			testing.assert(percentiles[percentile] !== false, 'Empty percentile for %s', percentile, callback);
		});

		testing.success(percentiles, callback);
	});
	for (var ms = 1; ms <= 10; ms++)
	{
		log.debug('Starting %s', ms);
		(function() {
			var id = latency.start();
			setTimeout(function()
			{
				log.debug('Ending %s', id);
				latency.end(id);
			}, ms);
		})();
	}
}

/**
 * A high resolution timer. Params:
 *	- delayMs: miliseconds to wait before calls. Can be fractional.
 *	- callback: function to call every time.
 */
exports.HighResolutionTimer = function(delayMs, callback)
{
	// self-reference
	var self = this;

	// attributes
	var counter = 0;
	var start = Date.now();
	var active = true;
	var timer;

	/**
	 * Delayed running of the callback.
	 */
	function delayed()
	{
		if (!active)
		{
			return false;
		}
		callback();
		counter ++;
		var diff = (Date.now() - start) - counter * delayMs;
		var timeout = Math.floor(delayMs - diff);
		if (timeout <= 0)
		{
			return delayed();
		}
		timer = setTimeout(delayed, delayMs - diff);
	}

	/**
	 * Show the drift of the timer.
	 */
	self.traceDrift = function()
	{
		var diff = Date.now() - start;
		var drift = diff / delayMs - counter;
		log.debug('Seconds: ' + Math.round(diff / 1000) + ', counter: ' + counter + ', drift: ' + drift);
	};

	/**
	 * Stop the timer.
	 */
	self.stop = function()
	{
		active = false;
		if (timer)
		{
			clearTimeout(timer);
			timer = null;
		}
	};

	// start timer
	delayed();
	timer = setTimeout(delayed, delayMs);
};

/**
 * Test a high resolution timer.
 */
function testTimer(callback)
{
	var timer = new exports.HighResolutionTimer(10, callback);
	timer.stop();
}

/**
 * Run package tests.
 */
exports.test = function(callback)
{
	var tests = [
		testLatencyIds,
		testLatencyRequests,
		testLatencyPercentiles,
		testTimer
	];
	testing.run(tests, callback);
};

// run tests if invoked directly
if (__filename == process.argv[1])
{
	exports.test(testing.show);
}

